include/grpc/credentials.h

```diff
@@ -1079,6 +1082,17 @@ GRPCAPI void grpc_tls_credentials_options_set_verify_server_cert(
 GRPCAPI void grpc_tls_credentials_options_set_send_client_ca_list(
     grpc_tls_credentials_options* options, bool send_client_ca_list);
 
+/**
+ * EXPERIMENTAL API - Subject to change
+ *
+ * Overrides the SNI that the client sends in the TLS handshake. nullopt
+ * indicates that SNI should not be overridden. An empty string value indicates
+ * that SNI should not be sent at all.
+ */
+GRPCAPI void grpc_tls_credentials_options_set_sni_override(
+    grpc_tls_credentials_options* options,
+    std::optional<std::string> sni_override);
```

Comment:
Although we have started to use C++ types in certain C-core APIs, this particular file hasn't done so yet.  I don't object to doing this, but let's be on the lookout for potential issues we may introduce in wrapped languages that may not be able to handle this yet.